### ISSUE TITLE
Nested modifications

### ISSUE THREAD

**Issue by fabiandev** [wix/tspoon#123]
Basically I want to transform the lowest level of a syntax kind first and have those changes reflected to the levels above. 

To achieve that, I've changed the order of AST processing, that the children's visit functions are called first.

Imagine you have a function with a body like below:

```ts
const foo = () => {
  const bar: string = 'test';
}
```

According to the AST, `foo` and `bar` is a `VariableDeclaration`.

I tried to process `bar` first, by replacing it with something else, which works fine.

But when `foo` is processed after that, the node passed to the visit function still reflects the old AST, without the changes made when `bar` was transformed.

Is it possible to have such nested transformations?

**Comment by tobich** [comment]
The changes caused by visitors are performed **after** all visitor calls, on all levels.

We are aware that the API, in its current form, may be misleading, creating impression that the changes are "imperative", and, therefore, immediate. We're discussing changing it towards more functional mode, where modifications are not "applied", but, rather, defined and returned as value.

**Comment by fabiandev** [comment]
I see, thanks for the clarification.

I'm currently looking forward to TypeScript releasing an official visitor/transformer API, like the PR https://github.com/Microsoft/TypeScript/pull/11561 suggests. This should happen soon, according to https://github.com/Microsoft/TypeScript/pull/11561#issuecomment-273984567.

Actually I built https://github.com/Microsoft/TypeScript/pull/11561/commits/6c122bcf16c329a21da04d7b5256f20ecc0bbd1d and I was able to achieve a transformation as I described above.

Thought this might be of interest to you for future versions of tspoon :)
